content/blog/2023/2024-01-27-deserialize-with-jsoncreator.md

```diff
@@ -0,0 +1,465 @@
+Subsequently, we will also take a look at a specific use case of using this annotation in a Spring Boot application. 
+For details on other commonly used Jackson annotations, refer to [this article](https://reflectoring.io/jackson/).
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/spring-boot/jackson-jsoncreator" %}}
+
+## What is @JsonCreator
+
+The @JsonCreator annotation is a part of the [Jackson API](https://fasterxml.github.io/jackson-annotations/javadoc/2.13/com/fasterxml/jackson/annotation/JsonCreator.html) that helps in deserialization. **Deserialization is a process of converting a JSON string into a Java object.**
+This is especially useful when we have multiple constructors/static factory methods for object creation. With the `@JsonCreator` annotation we can specify which constructor/static factory method to use during the deserialization process.
+
+## Working with @JsonCreator annotation
+
+In this section, we will look at a few use-cases of how this annotation works.
+### Deserializing immutable objects
+Java encourages creating immutable objects since they are thread-safe and easy to maintain. To get a better understanding of how to use immutable objects in java, refer to [this article.](https://reflectoring.io/java-immutables/)
+Let's try to deserialize an immutable object `UserData` which is defined as below :
+````java
+
+@JsonIgnoreProperties(ignoreUnknown=true)
+public class UserData {
+
+    private final long id;
+
+    private final String firstName;
+
+    private final String lastName;
+
+    private final String city;
+
+    private LocalDate createdDate;
+
+
+    public UserData(long id, String firstName, String lastName, String city) {
+        this.id = id;
+        this.firstName = firstName;
+        this.lastName = lastName;
+        this.city = city;
+        this.createdDate = LocalDate.now();
+    }
+
+    public UserData(long id, String firstName, String lastName, String city, 
+                    LocalDate createdDate) {
+        this.id = id;
+        this.firstName = firstName;
+        this.lastName = lastName;
+        this.city = city;
+        this.createdDate = createdDate;
+    }
+    
+    // Getters here...
+}
+````
+Next, let's try to deserialize the object with Jackson's **ObjectMapper** class:
+````java
+
+    @Test
+    public void deserializeImmutableObjects() throws JsonProcessingException {
+        String userData = 
+            objectMapper.writeValueAsString(MockedUsersUtility.getMockedUserData());
+        System.out.println("USER: " + userData);
+        UserData user = objectMapper.readValue(userData, UserData.class);
+        assertNotNull(user);
+    }
+    
+````
+In the above example the `writeValueAsString()` serializes the `UserData` object to a String.
+The `readValue()` method is responsible for deserializing the String Object back to the `UserData` object.
```

Comment:
Let's explicitly mention that the serialisation works. I know that it's obvious looking at the code and the error message, however, I think the article reads more fluently if we state it explicitly.